<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml" xmlns:th="http://www.thymeleaf.org">
<head>
    <meta charset="UTF-8">
    <title>SQUIRLS - Super-Quick Information Content and Random Forest Learning for Splice Variants</title>
    <meta content="width=device-width, initial-scale=1.0" name="viewport">
    <!-- TODO - Add the CSS into the template -->
    <!-- Latest compiled and minified CSS -->
    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css" rel="stylesheet">
    <style>
        .panel > .panel-heading {
            /*background-color: #00fa3c;*/
        }

        .panel > .panel-info {
            border-color: transparent;
            border-width: 0;
            margin: 5px 0 0 0;
        }

        .graphics-container {
            width: 100%;
            height: 100%;
            padding-top: 5px;
        }

        .graphics-title {
            text-align: center;
            text-transform: uppercase;
            font-weight: bold;
        }

        /* Individual figures are wrapped in a div of this class. */
        .graphics-subcontent {
            text-align: center;
            padding: 2rem;
            display: table-cell;
            vertical-align: middle;
        }

        .donor .graphics-subcontent:nth-child(1) svg:nth-child(1) {
            display: block;
        }

        .donor .graphics-content {
            margin: 0 auto;
            padding: 2rem;
            display: table;
        }

        /* All figures wrapped within `.graphics-subcontent` are wrapped within this element. */
        .graphics-content {
            max-width: available;
            padding: 2rem;
        }

        div svg {
            object-fit: contain;
        }

        ul.bottomspace {
            margin-bottom: 1cm;
        }

        .effects-title {
            text-transform: uppercase;
            font-weight: bold;
            margin: 5px 0 0 0;
        }

        .transcript-list {
            margin: 0 0 15px 0;
        }

        .feature-table tr {
            display: block;
            float: left;
        }

        .feature-table th, .feature-table td {
            display: block;
        }

        .feature-table td {
            text-align: center;
            border: solid .5px #ddd;
        }
    </style>
</head>
<body>
<div class="container" id="content">
    <!--    TITLE -->
    <h2>Results for <em th:text="${sampleName}">Sample ID</em></h2>

    <!--    NAV   -->
    <ul class="nav nav-pills" id="tabs">
        <li><a data-toggle="tab" href="#inputs" role="tab"><b>Analysis Inputs</b></a></li>
        <li><a data-toggle="tab" href="#statistics" role="tab"><b>Analysis Statistics</b></a></li>
        <li><a data-toggle="tab" href="#prioritisation" role="tab"><b>Prioritised Variants</b></a></li>
        <li><a data-toggle="tab" href="#about" role="tab"><b>About</b></a></li>
    </ul>

    <!--    INPUTS    -->
    <div class="panel panel-default" id="inputs">
        <div class="panel-heading">
            <h3 class="panel-title">
                Analysis input
            </h3>
        </div>
        <div class="panel-body">
            <p>Settings used for this analysis:</p>
            <pre th:text="${settings.getYamlRepresentation()}">
  - input VCF path: /path/to/input.vcf
  - Jannovar transcript database: /path/to/hg19_ensembl.ser
  - # variants reported: 100</pre>
        </div>
    </div>

    <!--    STATISTICS      -->
    <div class="panel panel-default" id="statistics">
        <div class="panel-heading">
            <h3 class="panel-title">Analysis statistics</h3>
        </div>
        <div class="panel-body">
            <table class="table table-striped">
                <tr>
                    <th>Info</th>
                    <th>Count</th>
                </tr>
                <tr>
                    <td>number of variants in the VCF file</td>
                    <td th:text="${stats.getAllVariants()}">100</td>
                </tr>
                <tr>
                    <td>number of <em>ALT</em> alleles in the VCF file</td>
                    <td th:text="${stats.alleleCount}">120</td>
                </tr>
                <tr>
                    <td>number of alleles with &gt;1 functional annotation</td>
                    <td th:text="${stats.annotatedAlleleCount}">110</td>
                </tr>
            </table>
        </div>
    </div>

    <!--    PRIORITISED VARIANTS     -->
    <div class="panel panel-default" id="prioritisation">
        <div class="panel-heading">
            <h3 class="panel-title">Prioritised Variants</h3>
        </div>
        <div class="panel panel-info" th:each="variant, variant_is: ${variants}">
            <div class="panel-heading">
                <div class="panel-title row">
                    <div class="col-md-4">
                        <b th:text="${variant.getVariantRepresentation()}">chr9:136,224,690A > G</b>
                    </div>
                    <div class="col-md-4">
                        <b><em th:text="${variant.getGeneSymbol()}">SURF2</em></b>
                    </div>
                    <div class="col-md-4">
                        Pathogenicity:
                        <b th:text="${#numbers.formatDecimal(variant.getMaxPathogenicity(), 1, 3)}">0.89</b>
                    </div>
                </div>
            </div>
            <div class="panel-body">
                <div th:remove="tag" th:utext="${variant.getContent()}">
                    <!-- CONTENT  -->
                </div>
            </div>
        </div>
    </div>

    <!--    FOOTER    -->
    <div class="panel panel-default" id="about">
        <div class="panel-heading">
            <h3 class="panel-title">About</h3>
        </div>
        <div class="panel-body">
            <p>
                <b>Squirls</b> is a desktop Java application that performs prediction of deleteriousness of genomic variants with respect to mRNA splicing.
                Variants are prioritized based on interpretable splicing features using a machine learning algorithm.
                The functional annotation code is based on <a href="https://github.com/charite/jannovar/">Jannovar</a>.
            </p>
            <!--todo Update documentation link-->
            <p>See <a href="#">the comprehensive documentation</a> for instructions regarding how to perform setup, run and interpret the results.</p>
            <p>Developed by <em>Daniel Danis</em> and <em>Peter N Robinson</em> at <a href="https://www.jax.org">The Jackson Laboratory for Genomic Medicine</a>.</p>
            <p>Problems, suggestions, or comments? Please <a href="mailto:daniel.danis@jax.org">let us know</a>.</p>

            <div style="text-align: center;">&#169;2020 Daniel Danis, Peter N Robinson</div>
        </div>
    </div>
</div>
</body>
</html>