<!DOCTYPE html>
<html lang="en" xmlns="http://www.w3.org/1999/xhtml" xmlns:th="http://www.thymeleaf.org">
<head>
    <meta charset="UTF-8">
    <title>SQUIRLS - Super-Quick Information Content and Random Forest Learning for Splice Variants</title>
    <meta content="width=device-width, initial-scale=1.0" name="viewport">
    <!-- TODO - Add the CSS into the template -->
    <!-- Latest compiled and minified CSS -->
    <link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css" rel="stylesheet">
    <style>
        .panel > .panel-heading {
            /*background-color: #00fa3c;*/
        }

        .panel > .panel-info {
            border-color: transparent;
            border-width: 0;
            margin: 5px 0 0 0;
        }

        .graphics-container {
            width: 100%;
            height: 100%;
            padding-top: 5px;
        }

        .graphics-title {
            text-align: center;
            text-transform: uppercase;
            font-weight: bold;
        }

        /* Individual figures are wrapped in a div of this class. */
        .graphics-subcontent {
            text-align: center;
            padding: 2rem;
            display: table-cell;
            vertical-align: middle;
        }

        .donor .graphics-subcontent:nth-child(1) svg:nth-child(1) {
            display: block;
        }

        .donor .graphics-content {
            margin: 0 auto;
            padding: 2rem;
            display: table;
        }

        /* All figures wrapped within `.graphics-subcontent` are wrapped within this element. */
        .graphics-content {
            max-width: available;
            padding: 2rem;
        }

        div svg {
            object-fit: contain;
        }

        ul.bottomspace {
            margin-bottom: 1cm;
        }

        .effects-title {
            text-transform: uppercase;
            font-weight: bold;
            margin: 5px 0 0 0;
        }

        .transcript-list {
            margin: 0 0 15px 0;
        }

        .feature-table tr {
            display: block;
            float: left;
        }

        .feature-table th, .feature-table td {
            display: block;
        }

        .feature-table td {
            text-align: center;
            border: solid .5px #ddd;
        }

        h4 {
            padding-top: .5rem;
            font-style: italic;
        }
    </style>
</head>
<body>
<div class="container" id="content">
    <!--    TITLE -->
    <h2><em>SQUIRLS</em> analysis results for <em th:text="${sampleName}">Sample ID</em></h2>

    <!--    NAV   -->
    <ul class="nav nav-pills" id="tabs">
        <li><a data-toggle="tab" href="#inputs" role="tab"><b>Analysis input</b></a></li>
        <li><a data-toggle="tab" href="#statistics" role="tab"><b>Statistics</b></a></li>
        <li><a data-toggle="tab" href="#prioritisation" role="tab"><b>Prioritised Variants</b></a></li>
        <li><a data-toggle="tab" href="#about" role="tab"><b>About</b></a></li>
    </ul>

    <!--    INPUTS    -->
    <div id="inputs">
        <h3>Analysis input</h3>
        <div>
            <b>Analyzed variants</b>
            <p>
                We analyzed VCF file with variants located at
            </p>
            <p><em><span th:remove="tag" th:text="${settings.getInputPath()}">/path/to/file.vcf</span></em></p>

            <b>Transcript database</b>
            <p>
                We use <a href="https://github.com/charite/jannovar/">Jannovar</a> library to perform functional annotation of variants
                <em>wrt.</em> genes and transcripts. The transcript definitions were read from the database file located at
            </p>
            <p><em><span th:text="${settings.getTranscriptDb()}">/path/to/jannovar.ser</span></em></p>

            <b>Number of reported variants</b>
            <p>
                After the analysis, the variants are sorted by predicted pathogenicity and the variant with the highest predicted
                pathogenicity is placed on the top of the list.
                This report presents the top <b><span th:text="${settings.getNReported()}">100</span></b> most pathogenic variants.
            </p>
        </div>
    </div>

    <!--    STATISTICS      -->
    <div id="statistics">
        <h3>Statistics</h3>
        <div>
            <table class="table table-striped">
                <tr>
                    <td>number of variants in the VCF file</td>
                    <td th:text="${stats.allVariants()}">100</td>
                </tr>
                <tr>
                    <td>number of <em>ALT</em> alleles in the VCF file</td>
                    <td th:text="${stats.alleleCount()}">120</td>
                </tr>
                <tr>
                    <td>number of <em>ALT</em> alleles annotated by Squirls</td>
                    <td th:text="${stats.annotatedAlleleCount()}">110</td>
                </tr>
                <tr>
                    <td>number of reported variants</td>
                    <td th:text="${variants.size()}">100</td>
                </tr>
            </table>
        </div>
    </div>

    <!--    PRIORITISED VARIANTS     -->
    <div id="prioritisation">
            <h3>Prioritised Variants</h3>
        <div class="panel panel-info" th:each="variant, variant_is: ${variants}">
            <div class="panel-heading">
                <div class="panel-title row">
                    <div class="col-md-4">
                        <b th:text="${variant.getVariantRepresentation()}">chr9:136,224,690A>G</b>
                    </div>
                    <div class="col-md-4">
                        <b><em th:text="${variant.getGeneSymbol()}">SURF2</em></b>
                    </div>
                    <div class="col-md-4">
                        Pathogenicity:
                        <b th:text="${#numbers.formatDecimal(variant.getMaxPathogenicity(), 1, 3)}">0.89</b>
                    </div>
                </div>
            </div>
            <div class="panel-body">
                <div th:remove="tag" th:utext="${variant.getContent()}">
                    <!-- CONTENT  -->
                </div>
            </div>
        </div>
    </div>

    <!--    FOOTER    -->
    <div id="about">
        <h3>About</h3>
        <div>
            <p>
                <em>SQUIRLS</em> is a desktop Java application that performs prediction of deleteriousness of genomic variants with respect to mRNA splicing.
                Variants are prioritized based on interpretable splicing features using a machine learning algorithm.
                The functional annotation code is based on <a href="https://github.com/charite/jannovar/">Jannovar</a>.
            </p>
            <p>See <a href="http://squirls.rtfd.io">the comprehensive documentation</a> for instructions regarding how to perform setup, run and interpret the results.</p>
            <p>Developed by <em>Daniel Danis</em> and <em>Peter N Robinson</em> at <a href="https://www.jax.org">The Jackson Laboratory for Genomic Medicine</a>.</p>
            <p>Problems, suggestions, or comments? Feel free to submit an issue to our <a href="https://github.com/TheJacksonLaboratory/Squirls/issues">GitHub tracker</a>.</p>

            <div style="text-align: center;">&#169; 2020 Daniel Danis, Peter N Robinson</div>
        </div>
    </div>
</div>
</body>
</html>