.. _rstoutputformats:

==============
Output formats
==============

Squirls supports storing results in 4 output formats: *HTML*, *VCF* *CSV*, and *TSV*. Use the ``-f`` option
to select one or more of the desired output formats (e.g. ``-f html,vcf``). *HTML* report is generated by default.

HTML output format
^^^^^^^^^^^^^^^^^^

Squirls creates an *HTML* file with the analysis summary and with variants sorted by Squirls score in descending order.
The number of the reported variants is adjusted by the ``-n`` option. See the :ref:`rstinterpretation` section
to learn more about the content of the *HTML* report.

VCF output format
^^^^^^^^^^^^^^^^^
When including ``vcf`` into the ``-f`` option, a VCF file with all input variants is created. The annotation process
adds a novel *FILTER* and *INFO* field(s) to each variant that overlaps with one or more transcript region:

* ``SQUIRLS`` - a *FILTER* flag indicating that the variant is considered to have a deleterious effect on >=1 overlapping transcript
* ``SQUIRLS_SCORE`` - an *INFO* string containing maximum SQUIRLS scores for the overlapping transcripts.
* ``SQUIRLS_TXS`` - an *INFO* string containing SQUIRLS scores for each  variant-transcript combination. The field is
  present when ``--all-transcripts`` option is specified.

For the example variant ``chr1:1234C>A,G``, the *INFO* field will contain the following::

    SQUIRLS_SCORE=0.988654;SQUIRLS_TXS=A|NM_123456.1=0.988654|ENST00000987654.1=0.988654
    SQUIRLS_SCORE=0.330112;SQUIRLS_TXS=G|NM_12356.1=0.330112|ENST00000987654.1=0.330112

Multi-allelic variants are broken down into separate records and processed individually. Predictions with respect to
the overlapping transcripts are separated by a pipe (``|``) symbol.

.. note::
  The ``-n`` option has no effect for the *VCF* output format.

.. note::
  The ``vcfgz`` output format has been deprecated. Use ``--compress`` and ``-f vcf`` options to get results as
  a compressed VCF file.

CSV/TSV output format
^^^^^^^^^^^^^^^^^^^^^
To write *n* most deleterious variants into a *CSV* (or *TSV*) file, use ``csv`` (``tsv``) in the ``-f`` option.

The results are written into a tabular file with the following columns:

.. table:: Tabular output

  ====== =========== ===== ===== ============= ============== ================ ================
  chrom   pos        ref   alt   gene_symbol   tx_accession    interpretation   squirls_score
  ====== =========== ===== ===== ============= ============== ================ ================
  chr3    165504107   A     C     *BCHE*        NM_000055.2    pathogenic       0.99997203304
  ...     ...         ...   ...   ...           ...            ...              ...
  ====== =========== ===== ===== ============= ============== ================ ================

.. note::
  Use ``--report-features`` and/or ``--all-transcripts`` options to add columns with Squirls features and/or Squirls
  scores for all overlapping transcripts.
